### PR TITLE
ALTAPPS-951: Shared parsons problem onboarding modal

### PR DESCRIPTION
**Pull request by vladkash**
**YouTrack Issues**:
[#ALTAPPS-951](https://vyahhi.myjetbrains.com/youtrack/issue/ALTAPPS-951)

**Checklist**

_Before Code Review:_

- [x] Fields "Assignees, Labels, Milestone" are filled in the pull request;
- [x] View analytics events are added for new screens;
- [x] New analytics events are documented;
- [x] All checks have been passed;

**Description**
- Added messages and actions for parsons problem onboarding modal to `StepQuizFeature`

### CODE REVIEW

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizReducer.kt [hunk 1]
```diff
@@ -249,6 +264,8 @@ class StepQuizReducer(
 
                 val actions = if (isProblemsLimitReached) {
                     setOf(Action.ViewAction.ShowProblemsLimitReachedModal)
+                } else if (message.shouldShowParsonsModal) {
+                    setOf(Action.ViewAction.ShowParsonsProblemOnboardingModal)
```
Comment (XanderZhu) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizReducer.kt:268]: Let's replace `else if` with `when`

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizReducer.kt [hunk 2]
```diff
@@ -228,6 +230,19 @@ class StepQuizReducer(
                         ProblemsLimitReachedModalHiddenHyperskillAnalyticEvent(stepRoute.analyticRoute)
                     )
                 )
+            is Message.ParsonsProblemOnboardingModalShownEventMessage ->
+                state to setOf(
+                    Action.LogAnalyticEvent(
+                        ParsonsProblemOnboardingModalShownHyperskillAnalyticEvent(stepRoute.analyticRoute)
+                    ),
+                    Action.SaveParsonsProblemOnboardingModalShownCacheFlag
+                )
```
Comment (XanderZhu) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizReducer.kt:239]: I suggest naming all the messages without the `Event` postfix to not expect some concrete behavior reading the name of the event.
In this case, `ParsonsProblemOnboardingModalShownEventMessage` suppose not only analytics sending.

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizActionDispatcher.kt [hunk 3]
```diff
@@ -53,10 +56,12 @@ class StepQuizActionDispatcher(
...
-                                        action.step,
-                                        attempt,
-                                        it,
-                                        isProblemsLimitReached
+                                        step = action.step,
+                                        attempt = attempt,
+                                        submissionState = it,
+                                        isProblemsLimitReached = isProblemsLimitReached,
+                                        shouldShowParsonsModal = action.step.block.name == BlockName.PARSONS &&
+                                            !onboardingInteractor.isParsonsOnboardingShown()
```
Comment (XanderZhu) [inline shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizActionDispatcher.kt:64]: Let's pass just a isParsonsOnboardingShown flag.
In my opinion, it's better to move step-type detection to reducer.

**Review by XanderZhu — APPROVED** [review]

File: shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizReducer.kt (hunk 2 above)
Comment (vladkash) [reply shared/src/commonMain/kotlin/org/hyperskill/app/step_quiz/presentation/StepQuizReducer.kt:239]: I agree, very good point :+1:

**Pull request merged**